apps/chat/conversation.py

```diff
@@ -192,69 +201,71 @@ def _compress_chat_history(
         len(total_messages),
         MAX_UNCOMPRESSED_MESSAGES,
     )
-
     history, last_message, summary = compress_chat_history_from_messages(
-        llm, history, keep_history_len, max_token_limit, input_messages
+        llm, history, keep_history_len, max_token_limit, input_messages, history_mode
     )
     return history, last_message, summary
 
+def truncate_tokens(history, max_token_limit, llm, input_message_tokens):
+    """Removes old messages until the token count is below the max limit."""
+    pruned_memory = []
+    history_tokens = llm.get_num_tokens_from_messages(history)
+    while history and history_tokens + input_message_tokens > max_token_limit:
+        pruned_memory.append(history.pop(0))
+        history_tokens = llm.get_num_tokens_from_messages(history)
+    return history, pruned_memory
 
-def compress_chat_history_from_messages(
-    llm, history, keep_history_len: int, max_token_limit: int, input_messages: list
-):
-    summary = history.pop(0).content if history[0].type == ChatMessageType.SYSTEM else None
-    history, pruned_memory = history[-keep_history_len:], history[:-keep_history_len]
-    # The latest message
-    latest_message = history[-1]
 
+def summarize_history(llm, history, max_token_limit, input_message_tokens, summary):
```

Comment:
seems like this implementation is missing some of the features of the existing implementation around the max messages length (_messages_exceeds_limit)